### PR TITLE
Convert TaintBasedEvictions e2e to integration test

### PR DESCRIPTION
**Pull request by damemi**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind cleanup

**What this PR does / why we need it**:
Rebases and updates the changes from https://github.com/kubernetes/kubernetes/pull/80817 with the feedback provided in that PR

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

### CODE REVIEW

**Comment by soltysh** [comment]
/sig scheduling
/kind cleanup
/priority important-longterm
@damemi you're missing bazel update

**Review by Huang-Wei — COMMENTED** [review]
Thanks, @damemi . Some comments below, and for reviewers' conveniences please make changes in a new commit, we can squash later.

**Review by Huang-Wei — COMMENTED** [review]
Some nits.

**Comment by damemi** [comment]
@Huang-Wei I know it's past code freeze now but I made some changes to this for you to consider. In my testing I ultimately had to move much of the code (everything after `initTestMaster`) back into the test iteration loop. Without initializing a new test context for each test, I ran into problems where the statuses applied to nodes during the first test affected following tests (resulting in errors like `Failed to taint node in test 1 <node-1>, err: timed out waiting for the condition`). Since everything that follows depends on the context from `initTestMaster`, I of course need to include that as well.

At this point the tests themselves should be passing, and I would appreciate your further review so that this can merge once master re-opens for 1.17

**Comment by Huang-Wei** [comment]
> Without initializing a new test context for each test, I ran into problems where the statuses applied to nodes during the first test affected following tests

That looks good to me.

Thanks, @damemi for your tremendous efforts on this. I will definitely follow up in the 1.17 release cycle.

**Comment by damemi** [comment]
@Huang-Wei with master reopened now, do you have a chance to look at this again?

File: test/integration/scheduler/taint_test.go [hunk 1]
```diff
@@ -571,3 +575,228 @@ func TestTaintNodeByCondition(t *testing.T) {
 		})
 	}
 }
+
+// TestTaintBasedEvictions tests related cases for the TaintBasedEvictions feature
+func TestTaintBasedEvictions(t *testing.T) {
```
Comment (oomichi) [inline test/integration/scheduler/taint_test.go]: Don't we need to remove e2e test `taint_based_evictions.go` according to the PR title?
The original PR (https://github.com/kubernetes/kubernetes/pull/80817) also removed the e2e test.

File: test/integration/scheduler/taint_test.go [hunk 2]
```diff
@@ -571,3 +575,228 @@ func TestTaintNodeByCondition(t *testing.T) {
...
+				{
+					Key:      schedulerapi.TaintNodeNotReady,
+					Operator: v1.TolerationOpExists,
+					Effect:   v1.TaintEffectNoExecute,
+				},
+			},
+			TerminationGracePeriodSeconds: &gracePeriod,
+		},
+	}
+	tolerationSeconds := []int64{200, 300, 0}
```
Comment (oomichi) [inline test/integration/scheduler/taint_test.go]: How about defining `tolerationSeconds` in each test struct like
```
		{
			name:                "Taint based evictions for NodeNotReady and 200 tolerationseconds",
			nodeTaints:          []v1.Taint{{Key: schedulerapi.TaintNodeNotReady, Effect: v1.TaintEffectNoExecute}},
			nodeConditions:      []v1.NodeCondition{{Type: v1.NodeReady, Status: v1.ConditionFalse}},
			pod:                 testPod,
			waitForPodCondition: "updated with tolerationSeconds of 200",
			tolerationSeconds:   200,
```
for the readability?

Comment (damemi) [reply test/integration/scheduler/taint_test.go]: TolerationSeconds needs to be a pointer value https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2814

So, we need to declare it in a variable somewhere (can't just do a pointer to an int literal) so we decided this was the way to do it (see https://github.com/kubernetes/kubernetes/pull/81856#discussion_r317727639)

File: test/integration/scheduler/taint_test.go (hunk 1 above)
Comment (damemi) [reply test/integration/scheduler/taint_test.go]: done, looks like that got lost in some of my rebasing

File: test/integration/scheduler/taint_test.go [hunk 3]
```diff
@@ -571,3 +575,228 @@ func TestTaintNodeByCondition(t *testing.T) {
...
+	// Enable TaintBasedEvictions
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.TaintBasedEvictions, true)()
+
+	// Build admission chain handler.
+	podTolerations := podtolerationrestriction.NewPodTolerationsPlugin(&pluginapi.Configuration{})
+	admission := admission.NewChainHandler(
+		podTolerations,
+		defaulttolerationseconds.NewDefaultTolerationSeconds(),
+	)
+	for i, test := range tests {
```
Comment (Huang-Wei) [inline test/integration/scheduler/taint_test.go]: We'd better follow Golang subtest standard here:

```go
	for i, test := range tests {
		t.Run(test.name, func(t *testing.T) {
```

File: test/integration/scheduler/taint_test.go [hunk 4]
```diff
@@ -571,3 +575,228 @@ func TestTaintNodeByCondition(t *testing.T) {
...
+		// Build clientset and informers for controllers.
+		externalClientset := kubernetes.NewForConfigOrDie(&restclient.Config{
+			QPS:           -1,
+			Host:          context.httpServer.URL,
+			ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
+		externalInformers := informers.NewSharedInformerFactory(externalClientset, time.Second)
+		podTolerations.SetExternalKubeClientSet(externalClientset)
+		podTolerations.SetExternalKubeInformerFactory(externalInformers)
+		// ApplyFeatureGates() is called to ensure TaintNodesByCondition related logic is applied/restored properly.
+		defer algorithmprovider.ApplyFeatureGates()()
```
Comment (Huang-Wei) [inline test/integration/scheduler/taint_test.go]: Do you mind moving this below L645?

**Comment by damemi** [comment]
@Huang-Wei pushed your suggestions in a new commit. If this looks good let me know and I will squash

File: test/integration/scheduler/taint_test.go [hunk 5]
```diff
@@ -571,3 +575,230 @@ func TestTaintNodeByCondition(t *testing.T) {
...
+				100,              // Secondary eviction limiter QPS
+				50,               // Large cluster threshold
+				0.55,             // Unhealthy zone threshold
+				true,             // Run taint manager
+				true,             // Use taint based evictions
+				false,            // Enabled TaintNodeByCondition feature
+			)
+			if err != nil {
+				t.Errorf("Failed to create node controller: %v", err)
+				return
```
Comment (oomichi) [inline test/integration/scheduler/taint_test.go]: nit: This `return` might be unnecessary.

**Review by oomichi — COMMENTED** [review]
Thanks for updating,

/approve

**Comment by damemi** [comment]
Since this has an /approve, I'm going to squash

**Comment by Huang-Wei** [comment]
/lgtm

Thanks @damemi for the efforts!

**Pull request merged**